### PR TITLE
Feature/rpnv1 507 deterministic test epochs

### PR DESCRIPTION
**Pull request by clstrfsck**

### CODE REVIEW

File: radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/RandomChannelOrderResponsiveTest.java [hunk 1]
```diff
@@ -17,32 +17,64 @@
...
+		final int numNodes = 100;
+		final int viewsToRun = 50 * numNodes; // numNodes should be a factor of this number
 		final Random random = new Random(12345);
-		final DeterministicTest test = DeterministicTest.createSingleEpochAlwaysSyncedTest(100);
 
-		test.start();
-		for (int step = 0; step < NUM_STEPS; step++) {
-			test.processNextMsg(random);
-		}
+		DeterministicTest test = DeterministicTest.builder()
```
Comment (talekhinezh) [inline radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/RandomChannelOrderResponsiveTest.java]: Would be good for both tests in this class. to share the same builder and only configure the numNodes.

File: radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/OneSlowNodeTest.java [hunk 2]
```diff
@@ -30,24 +43,67 @@
...
+			final Class<?> expectedMsgClass = messageDetails.getSecond();
+			for (ControlledMessage message : messages) {
+				if (expectedChannel.equals(message.channelId())) {
+					Class<?> msgClass = message.message().getClass();
+					if (expectedMsgClass.isAssignableFrom(msgClass)) {
+						return message;
+					}
+				}
+			}
+			fail(String.format("Can't find %s message %s: %s", expectedMsgClass.getSimpleName(), expectedChannel, messages));
```
Comment (talekhinezh) [inline radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/OneSlowNodeTest.java]: why not just throw an exception?

File: radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/OneSlowNodeTest.java [hunk 3]
```diff
@@ -30,24 +43,67 @@
...
+		DeterministicTest.builder()
+			.numNodes(numNodes)
+			.syncedExecutorFactory(SyncedExecutorFactories.alwaysSynced())
+			.messageSelector(sequenceSelector(expectedSequence))
+			.messageMutator(stopWhenEmpty(expectedSequence).otherwise(delayMessagesForNode(0)))
+			.build()
+			.run();
+	}
+
+	private MessageSelector sequenceSelector(LinkedList<Pair<ChannelId, Class<?>>> expectedSequence) {
```
Comment (talekhinezh) [inline radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/OneSlowNodeTest.java]: Should be static

File: radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/OneSlowNodeTest.java [hunk 4]
```diff
@@ -30,24 +43,67 @@
...
+						return message;
+					}
+				}
+			}
+			fail(String.format("Can't find %s message %s: %s", expectedMsgClass.getSimpleName(), expectedChannel, messages));
+			return null; // Not required, but compiler can't tell that fail throws exception
+		};
+	}
+
+	private MessageMutator stopWhenEmpty(Collection<?> expectedSequence) {
```
Comment (talekhinezh) [inline radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/OneSlowNodeTest.java]: Should be static

File: radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/OneSlowNodeTest.java [hunk 5]
```diff
@@ -30,24 +43,67 @@
...
+		return (rank, message, queue) -> !expectedSequence.isEmpty();
+	}
 
-		test.processNextMsg(0, 2, Proposal.class);
+	private MessageMutator delayMessagesForNode(int index) {
```
Comment (talekhinezh) [inline radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/OneSlowNodeTest.java]: Should be static

File: radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/network/MessageSelector.java [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+import java.util.List;
+import java.util.Random;
+
+/**
+ * Select a message from a list of possible messages.
+ */
+@FunctionalInterface
+public interface MessageSelector {
+	ControlledMessage select(List<ControlledMessage> messages);
```
Comment (talekhinezh) [inline radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/network/MessageSelector.java:42]: Does the list order mean anything? Is it in the order of messages as they arrived?

File: radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/DifferentTimestampsCauseTimeoutTest.java [hunk 7]
```diff
@@ -146,4 +214,29 @@ private TimestampedECDSASignature mutateTimestampedSignature(TimestampedECDSASig
...
+			final Class<?> expectedMsgClass = messageDetails.getSecond();
+			for (ControlledMessage message : messages) {
+				if (expectedChannel.equals(message.channelId())) {
+					Class<?> msgClass = message.message().getClass();
+					if (expectedMsgClass.isAssignableFrom(msgClass)) {
+						return message;
+					}
+				}
+			}
+			fail(String.format("Can't find %s message %s: %s", expectedMsgClass.getSimpleName(), expectedChannel, messages));
```
Comment (talekhinezh) [inline radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/DifferentTimestampsCauseTimeoutTest.java]: Why not exception? Would like these tests to move away from JUnit framework at some point

File: radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/OneSlowNodeTest.java [hunk 8]
```diff
@@ -30,24 +43,67 @@
 	 */
 	@Test
 	public void when_three_fast_nodes_and_one_slow_node_two_cycles__then_missing_parent_should_not_cause_sync_exception() {
-		final DeterministicTest test = DeterministicTest.createSingleEpochAlwaysSyncedTest(4);
-
-		test.start();
+		final int numNodes = 4;
 
+		LinkedList<Pair<ChannelId, Class<?>>> expectedSequence = Lists.newLinkedList();
```
Comment (talekhinezh) [inline radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/OneSlowNodeTest.java]: "expectedSequence" seems to be the wrong term as it's not "expected" but rather controlled

File: radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/network/MessageMutator.java [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+import com.radixdlt.consensus.bft.View;
+import com.radixdlt.consensus.epoch.EpochView;
+
+/**
+ * A mutator called on new messages that can mutate the message, or the queue.
+ * TODO: Could be broken up into a pipeline of filters and a single mutator.
+ */
+@FunctionalInterface
+public interface MessageMutator {
+	boolean mutate(MessageRank rank, ControlledMessage message, MessageQueue queue);
```
Comment (talekhinezh) [inline radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/network/MessageMutator.java:39]: Seems to me that this interface is conflating both being a message communication adversary and deciding how long a test should run. Think this should either change or add a comment that this needs to be split out.

File: radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/OneProposalTimeoutResponsiveTest.java [hunk 10]
```diff
@@ -61,22 +57,21 @@ private void run(int numNodes, long numViews, long dropPeriod) {
...
-			if (nv.getView().number() > numViews) {
-				this.nodesCompleted.add(senderId);
-				return false;
-			}
-		}
-		if (msg instanceof Proposal) {
-			final Proposal proposal = (Proposal) msg;
-			final View view = proposal.getVertex().getView();
-			final long viewNumber = view.number();
+	private MessageMutator dropSomeProposals(long dropPeriod) {
```
Comment (talekhinezh) [inline radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/OneProposalTimeoutResponsiveTest.java]: Should be static

File: radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/RandomChannelOrderResponsiveTest.java [hunk 11]
```diff
@@ -17,32 +17,64 @@
...
+			.numNodes(numNodes)
+			.syncedExecutorFactory(SyncedExecutorFactories.alwaysSynced())
+			.messageSelector(MessageSelector.randomSelector(random))
+			.messageMutator(MessageMutator.alwaysAddUntilView(View.of(viewsToRun)))
+			.build()
+			.run();
```
Comment (talekhinezh) [inline radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/RandomChannelOrderResponsiveTest.java:49]: Are timeouts disabled?

File: radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/RandomChannelOrderResponsiveTest.java [hunk 12]
```diff
@@ -17,32 +17,64 @@
...
```
Comment (talekhinezh) [inline radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/RandomChannelOrderResponsiveTest.java]: Are timeouts disabled?

File: radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/OneProposalDropperResponsiveTest.java [hunk 13]
```diff
@@ -32,30 +36,38 @@
...
+		DeterministicTest.builder()
+			.numNodes(numNodes)
+			.syncedExecutorFactory(SyncedExecutorFactories.alwaysSynced())
+			.messageSelector(MessageSelector.randomSelector(random))
+			.messageMutator(MessageMutator.stopAt(NUM_STEPS).otherwise(dropNode(numNodes, nodeToDropFunction)))
+			.build()
+			.run();
+	}
+
+	private MessageMutator dropNode(int numNodes, Function<View, Integer> nodeToDropFunction) {
```
Comment (talekhinezh) [inline radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/OneProposalDropperResponsiveTest.java:47]: Should be static

File: radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/FProposalDropperResponsiveTest.java [hunk 14]
```diff
@@ -21,64 +21,72 @@
...
+		DeterministicTest.builder()
+			.numNodes(numNodes)
+			.syncedExecutorFactory(SyncedExecutorFactories.alwaysSynced())
+			.messageSelector(MessageSelector.randomSelector(random))
+			.messageMutator(MessageMutator.stopAt(NUM_STEPS).otherwise(dropNodes(numNodes, nodesToDropFunction)))
+			.build()
+			.run();
+	}
+
+	private MessageMutator dropNodes(int numNodes, Function<View, Set<Integer>> nodesToDropFunction) {
```
Comment (talekhinezh) [inline radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/FProposalDropperResponsiveTest.java]: Should be static

File: radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/FProposalDropperRandomSyncResponsiveTest.java [hunk 15]
```diff
@@ -21,64 +21,72 @@
...
+		DeterministicTest.builder()
+			.numNodes(numNodes)
+			.syncedExecutorFactory(SyncedExecutorFactories.randomlySynced(random))
+			.messageSelector(MessageSelector.randomSelector(random))
+			.messageMutator(MessageMutator.stopAt(NUM_STEPS).otherwise(dropNodes(numNodes, nodesToDropFunction)))
+			.build()
+			.run();
+	}
+
+	private MessageMutator dropNodes(int numNodes, Function<View, Set<Integer>> nodesToDropFunction) {
```
Comment (talekhinezh) [inline radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/FProposalDropperRandomSyncResponsiveTest.java]: Should be static

File: radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/network/ControlledSender.java [hunk 16]
```diff
@@ -0,0 +1,190 @@
...
+		public String toString() {
+			return String.format("%s{count=%s}", this.getClass().getSimpleName(), count);
+		}
+	}
+
+	private void handleMessage(MessageRank eav, ControlledMessage controlledMessage) {
+		this.network.handleMessage(eav, controlledMessage);
+	}
+
+	private MessageRank messageRank(EpochChange epochChange) {
```
Comment (talekhinezh) [inline radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/network/ControlledSender.java:176]: Seems to me that MessageRank shouldn't be part of ControlledSender as it's more of a communication adversary role. But fine for now but perhaps add a comment that this should move out of this class.

File: radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/DeterministicSyncExecutionModule.java [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+import com.radixdlt.integration.distributed.deterministic.configuration.SyncedExecutorFactory;
+import com.radixdlt.middleware2.CommittedAtom;
+import com.radixdlt.syncer.EpochChangeSender;
+import com.radixdlt.syncer.SyncExecutor.CommittedStateSyncSender;
+
+public class DeterministicSyncExecutionModule extends AbstractModule {
+	private final BFTValidatorSet validatorSet;
+	private final SyncedExecutorFactory syncedFactory;
+
+	public DeterministicSyncExecutionModule(BFTValidatorSet validatorSet, SyncedExecutorFactory syncedFactory) {
```
Comment (talekhinezh) [inline radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/DeterministicSyncExecutionModule.java]: factory in a module definitely smelly but fine for now, perhaps just a comment saying this should be fixed.

**Review by talekhinezh — CHANGES_REQUESTED** [review]

File: radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/OneSlowNodeTest.java (hunk 2 above)
Comment (clstrfsck) [reply radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/OneSlowNodeTest.java]: Could do that, but this is part of the test — failure seems more appropriate than an error here.

File: radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/OneSlowNodeTest.java (hunk 3 above)
Comment (clstrfsck) [reply radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/OneSlowNodeTest.java]: OK, done.

File: radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/OneSlowNodeTest.java (hunk 4 above)
Comment (clstrfsck) [reply radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/OneSlowNodeTest.java]: OK, gone.

File: radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/OneSlowNodeTest.java (hunk 5 above)
Comment (clstrfsck) [reply radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/OneSlowNodeTest.java]: OK, done.

File: radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/FProposalDropperResponsiveTest.java (hunk 14 above)
Comment (clstrfsck) [reply radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/FProposalDropperResponsiveTest.java]: OK, done.

File: radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/FProposalDropperRandomSyncResponsiveTest.java (hunk 15 above)
Comment (clstrfsck) [reply radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/FProposalDropperRandomSyncResponsiveTest.java]: OK, done.

File: radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/network/ControlledSender.java (hunk 16 above)
Comment (clstrfsck) [reply radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/network/ControlledSender.java:176]: These are all defaults, and message mutators can change them as they see fit, so leaving this as-is.

File: radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/OneProposalDropperResponsiveTest.java (hunk 13 above)
Comment (clstrfsck) [reply radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/OneProposalDropperResponsiveTest.java:47]: OK, done.

File: radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/DeterministicSyncExecutionModule.java (hunk 17 above)
Comment (clstrfsck) [reply radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/DeterministicSyncExecutionModule.java]: ~~Actually, I don't think this is required to be like this, so will change.~~

~~Does need to be here for now for the randomly synced executor to have its CommittedStateSyncSender injected, but I think the way that this happens needs to change in the future in any case, so leaving for now.~~

Actually it has changed now, and this module has now been replaced with DeterministicRandomlySyncedExecutorModule.

File: radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/network/MessageSelector.java (hunk 6 above)
Comment (clstrfsck) [reply radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/network/MessageSelector.java:42]: Yes, have improved the documentation here.

File: radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/DifferentTimestampsCauseTimeoutTest.java (hunk 7 above)
Comment (clstrfsck) [reply radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/DifferentTimestampsCauseTimeoutTest.java]: We can do that at the time that happens.  For now, this is a test failure, not an error, so this seems to be the best option with what we are currently working with.

File: radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/OneSlowNodeTest.java (hunk 8 above)
Comment (clstrfsck) [reply radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/OneSlowNodeTest.java]: Hmm.  Not really.  They are the messages that we expect to see in the queue, together with the order that we want to see them.

File: radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/network/MessageMutator.java (hunk 9 above)
Comment (clstrfsck) [reply radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/network/MessageMutator.java:39]: Yeah, this was what I was talking about yesterday.  It's changed.

File: radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/OneProposalTimeoutResponsiveTest.java (hunk 10 above)
Comment (clstrfsck) [reply radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/OneProposalTimeoutResponsiveTest.java]: OK, done.

File: radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/RandomChannelOrderResponsiveTest.java (hunk 11 above)
Comment (clstrfsck) [reply radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/RandomChannelOrderResponsiveTest.java:49]: Not here, but that has been fixed too.

File: radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/RandomChannelOrderResponsiveTest.java (hunk 12 above)
Comment (clstrfsck) [reply radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/RandomChannelOrderResponsiveTest.java]: Not in this code, but fixed later.

File: radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/RandomChannelOrderResponsiveTest.java (hunk 1 above)
Comment (clstrfsck) [reply radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/RandomChannelOrderResponsiveTest.java]: OK, done.

File: radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/OneSlowNodeTest.java (hunk 8 above)
Comment (talekhinezh) [reply radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/OneSlowNodeTest.java]: If so, then `sequenceSelector` should be changed to reflect this:
```
                        for (ControlledMessage message : messages) {
				if (expectedChannel.equals(message.channelId())) {
					Class<?> msgClass = message.message().getClass();
					if (expectedMsgClass.isAssignableFrom(msgClass)) {
						return message;
					} else {
                                                // ERROR
                                        }
				}
			}
```

File: radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/network/ControlledSender.java (hunk 16 above)
Comment (clstrfsck) [reply radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/network/ControlledSender.java:176]: Just to follow up on this, message ranking is required for timeouts to work properly.  The processing for timeouts needs to be delayed sufficiently to allow consensus rounds to complete before the timeout message is processed.

File: radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/OneSlowNodeTest.java (hunk 8 above)
Comment (clstrfsck) [reply radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/tests/consensus/OneSlowNodeTest.java]: Hmm, maybe I didn't say that correctly.  These are the messages we expect to see in the queue, together with the order that we want to process them in.

In particular, the test is implemented such that it delays processing the messages from node 0.  This didn't change from the original implementation, and your proposed implementation won't work for this test.

Anyhow, I'll change it to processingSequence.

File: radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/network/ControlledSender.java (hunk 16 above)
Comment (talekhinezh) [reply radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/network/ControlledSender.java:176]: right makes sense, but still there is no "default" to the timeout, its simply whatever the communication adversary is. But what we have here is fine for now.

**Review by talekhinezh — APPROVED** [review]

**Pull request merged**